### ISSUE TITLE
Several issues in vscode with the bundle verison

### ISSUE THREAD

**Issue by GilShoshan94** [tamasfe/taplo#280]
Hi,

Thank you for this library cli and extension.
I am using [the extension in vscode "Even Better TOML"](https://marketplace.visualstudio.com/items?itemName=tamasfe.even-better-toml). (in bundle mode)

I am on Windows 10 in case that matters.

I found several issues:

1) The config set in the setting.json of vscode does not "hot reload". 
I mean that when we change a setting, like setting `"evenBetterToml.formatter.indentEntries"` to `true` and go back to a toml file, the change does not apply, I found out that by reloading vscode (`ctrl+shift+p` : `Reload window`), only than the change apply and the toml file get correctly format as set in the settings.

2) I couldn't get [the configuration file `taplo.toml`](https://taplo.tamasfe.dev/configuration/file.html) to work. 
Even when specifically pointing to it in the `settings.json` with `"evenBetterToml.taplo.configFile.path": "taplo.toml"` and by reloading vscode. It is just ignore it seems to me.

---
For point 1, this is not really an issue, but if left as is, I think it should be mention in the documentation, as usually the UX with the other extensions in vscode is to apply immediately the new settings without a manual reload.

---
For point 2, let me know if what I can provide/test to help you figure it out.

**Comment by tamasfe** [comment]
I've created #282 for 1.

As for 2., I suspect it's a windows-specific issue, file paths are quite weird and currently no extra steps are taken in order to handle them. If this is the case, unfortunately I don't think you can help a lot without knowing the code base and I will have to get a windows machine.

The language server logs might have something interesting though, could you paste them here? (`OUTPUT` panel, and then select `Even Better TOML LSP` in the dropdown).

**Comment by GilShoshan94** [comment]
Sure.

This is all what I got printed there:

```
 INFO registered request handler method="initialize"
 INFO registered request handler method="textDocument/foldingRange"
 INFO registered request handler method="textDocument/documentSymbol"
 INFO registered request handler method="textDocument/formatting"
 INFO registered request handler method="textDocument/completion"
 INFO registered request handler method="textDocument/hover"
 INFO registered request handler method="textDocument/documentLink"
 INFO registered request handler method="textDocument/semanticTokens/full"
 INFO registered request handler method="textDocument/prepareRename"
 INFO registered request handler method="textDocument/rename"
 INFO registered notification handler method="initialized"
 INFO registered notification handler method="textDocument/didOpen"
 INFO registered notification handler method="textDocument/didChange"
 INFO registered notification handler method="textDocument/didSave"
 INFO registered notification handler method="textDocument/didClose"
 INFO registered notification handler method="workspace/didChangeConfiguration"
 INFO registered notification handler method="workspace/didChangeWorkspaceFolders"
 INFO registered request handler method="taplo/convertToJson"
 INFO registered request handler method="taplo/convertToToml"
 INFO registered request handler method="taplo/listSchemas"
 INFO registered request handler method="taplo/associatedSchema"
 INFO registered notification handler method="taplo/associateSchema"

```

**Comment by GilShoshan94** [comment]
Also, whenever I change a setting related to `evenBetterToml` in `settings.json` and save, I get this line printed in the `Even Better TOML LSP`'s `OUTPUT`:

```
 WARN no notification handler registered method=$/setTrace
```

**Comment by tamasfe** [comment]
Thanks, unfortunately not too much in the logs about it, but this is apparently 2 issues in one:

- The language server did not use the specified config path, whoops. I've just pushed a fix for this.
- The automatic discovery in the workspace still doesn't work most likely due to the path handling.

**Comment by GilShoshan94** [comment]
Is there a way to print out the logs from tracing in the `Even Better TOML LSP`'s `OUTPUT` ?
Maybe with a additional config like `"evenBetterToml.debugOutput": true` ?
Or by setting an environment variable with the existing setting `"evenBetterToml.taplo.environment": {},` ?

So users (like me) can help more to test and debug on various platform by providing the trace info.

**Comment by tamasfe** [comment]
It is possible to use the `RUST_LOG` environment variable to enable more verbose logs:

The following configuration will output significantly more logs:

```
  "evenBetterToml.taplo.environment": {
    "RUST_LOG": "info,taplo=debug"
  }
```

The environment variables are not passed to the bundled language server right now, I'll publish a new extension version that has this fixed along with the config file fix above soon.

To log pretty much everything you can use following settings, although it will be very noisy:

```
  "evenBetterToml.taplo.environment": {
    "RUST_LOG": "trace"
  },
  "evenBetterToml.taplo.extraArgs": ["--log-spans", "--verbose"],
```

Extra arguments will only work with a native `taplo` executable installed as the bundled server is just the language server and not a full `taplo-cli`.

**Comment by tamasfe** [comment]
The new version has just been published, can you please use the above settings and look for a line similar to this one:

```
DEBUG initialize:initialize: discovering config file in workspace self.root=file:///home/tamasfe/work/opensauce/taplo
```

`self.root` is the workspace root where the language server is looking for the configuration file.

**Comment by GilShoshan94** [comment]
I updated and set the config as you said.

The output is very verbose as you warned.
I found this in :
```
DEBUG handle_request: request received id=Number(0) method=initialize inner=Mutex { is_locked: false, has_waiters: false } 
request=Request { jsonrpc: "2.0", method: "initialize", id: Some(Number(0)), params: Some(Object({"processId": Number(5572), 
"clientInfo": Object({"name": String("Visual Studio Code"), 
"version": String("1.69.1")}), 
"locale": String("en"), 
"rootPath": String("c:\\Users\\PC-Gil\\Documents\\Work\\RustWorkspace\\nova"), 
"rootUri": String("file:///c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova"), 
"capabilities": Object({"workspace": Object({"applyEdit": Bool(true), 
"workspaceEdit": Object({"documentChanges": Bool(true), "resourceOperations": Array([String("create"), String("rename"), String("delete")]), "failureHandling": String("textOnlyTransactional"), "normalizesLineEndings": Bool(true), "changeAnnotationSupport": Object({"group .....
....
```

I reformatted a little to make it clearer.
The `"rootPath"` and `"rootUri"` are correct.

---

I searched for the `initialize:initialize`, got 464 results, the first:
```
DEBUG handle_request:request handler:initialize:initialize: discovering config file in workspace inner=Mutex 
{ is_locked: false, has_waiters: false } request=Request { jsonrpc: "2.0", 
method: "initialize", id: Some(Number(0)), params: Some(Object({"processId": Number(5572), 
"clientInfo": Object({"name": String("Visual Studio Code"), 
"version": String("1.69.1")}), 
"locale": String("en"), 
"rootPath": String("c:\\Users\\PC-Gil\\Documents\\Work\\RustWorkspace\\nova"), 
"rootUri": String("file:///c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova"), 
"capabilities": ...
....
```

and then:

```
DEBUG handle_request:request handler:initialize:initialize: built glob set; 0 literals, 3 basenames, 0 extensions, 0 prefixes, 0 suffixes, 0 required extensions, 0 regexes     inner=Mutex { is_locked: false, has_waiters: false } request=Request { ....
```
after the `request=Request { ` it seems to be always the same.

I then search for the config file `taplo.toml` with is in "rootPath" ( "rootPath"/taplo.toml) but found no result.

I looked around more and found this:
```
DEBUG update_configuration:initialize: built glob set; 0 literals, 0 basenames, 0 extensions, 0 prefixes, 0 suffixes, 2 required extensions, 0 regexes     self.root=file:///c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova
DEBUG update_configuration:initialize: built glob set; 1 literals, 0 basenames, 0 extensions, 0 prefixes, 0 suffixes, 0 required extensions, 0 regexes     self.root=file:///c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova
DEBUG update_configuration:initialize: built glob set; 1 literals, 0 basenames, 0 extensions, 0 prefixes, 0 suffixes, 2 required extensions, 0 regexes     self.root=file:///c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova
DEBUG update_configuration:initialize: using config file at specified path path="" self.root=file:///c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova
 INFO update_configuration:initialize: using config file path="/c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova/" self.root=file:///c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova
 WARN update_configuration:initialize: failed to load workspace configuration error=JsValue(Error: ENOENT: no such file or directory, open 'c:\c%3A\Users\PC-Gil\Documents\Work\RustWorkspace\nova'
	Error: ENOENT: no such file or directory, open 'c:\c%3A\Users\PC-Gil\Documents\Work\RustWorkspace\nova') self.root=file:///c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova
DEBUG update_configuration:initialize: built glob set; 5 literals, 0 basenames, 0 extensions, 0 prefixes, 0 suffixes, 0 required extensions, 0 regexes     self.root=file:///c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova
```


Does it help you ?

**Comment by GilShoshan94** [comment]
I added in `settings.json`
```json
"evenBetterToml.taplo.configFile.path": "taplo.toml",
```

and now when I searched for "taplo.toml" in the trace, this is what I can find:
```
...
DEBUG update_configuration:initialize: using config file at specified path path="taplo.toml" self.root=root:///
DEBUG update_configuration:initialize: relative config path is not valid for detached workspace self.root=root:///
...
DEBUG update_configuration:initialize: using config file at specified path path="taplo.toml" self.root=file:///c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova
 INFO update_configuration:initialize: using config file path="/c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova/taplo.toml" self.root=file:///c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova
 WARN update_configuration:initialize: failed to load workspace configuration error=JsValue(Error: ENOENT: no such file or directory, open 'c:\c%3A\Users\PC-Gil\Documents\Work\RustWorkspace\nova\taplo.toml'
	Error: ENOENT: no such file or directory, open 'c:\c%3A\Users\PC-Gil\Documents\Work\RustWorkspace\nova\taplo.toml') self.root=file:///c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova
```

which is weird because it seems like it found the file. 
"/c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova/taplo.toml"...

![image](https://user-images.githubusercontent.com/29143941/179405075-6c2723bc-e62a-4dbf-9ea3-1aa25a71bdf3.png)

**Comment by tamasfe** [comment]
It did help, thank you!

The issue is with `c%3A/`, it is `c:` that had to be URL-encoded since file paths are in URLs and they cannot contain `:` anywhere other than in the scheme, we currently don't decode any URLs sent by vscode.

I believe currently the only way to make this work is setting an absolute path. But you'll likely run into more issues until proper windows path handling is implemented.

**Comment by GilShoshan94** [comment]
Ok I found something else:
I set the path in `settings.json` explicitly this time, as such:
```
"evenBetterToml.taplo.configFile.path": "C:\\Users\\PC-Gil\\Documents\\Work\\RustWorkspace\\nova\\taplo.toml",
```

Now I can see in the trace that it found the config file, read it correctly, but does not apply it, or at least does no override the config set in `settings.json` (I set explicitly all 16 "evenBetterToml.formatter.*" settings).

```
...
DEBUG update_configuration:initialize: using config file at specified path path="C:\\Users\\PC-Gil\\Documents\\Work_Neurosteer\\RustWorkspace\\nova\\taplo.toml" self.root=root:///
 INFO update_configuration:initialize: using config file path="C:\\Users\\PC-Gil\\Documents\\Work\\RustWorkspace\\nova\\taplo.toml" self.root=root:///
DEBUG update_configuration:initialize: built glob set; 0 literals, 0 basenames, 1 extensions, 0 prefixes, 0 suffixes, 0 required extensions, 0 regexes     self.root=root:///
DEBUG update_configuration:initialize: using config: Config {
    include: None,
    exclude: None,
    rule: [],
    global_options: Options {
        schema: None,
        formatting: Some(
            OptionsIncomplete {
                align_entries: Some(
                    false,
                ),
                align_comments: None,
                array_trailing_comma: None,
                array_auto_expand: None,
                array_auto_collapse: None,
                compact_arrays: None,
                compact_inline_tables: None,
                compact_entries: None,
                column_width: None,
                indent_tables: None,
                indent_entries: None,
                indent_string: None,
                trailing_newline: None,
                reorder_keys: Some(
                    true,
                ),
                allowed_blank_lines: None,
                crlf: None,
            },
        ),
    },
} self.root=root:///
...
```

Which is correct I this is the content of my `taplo.toml`:
```toml
[formatting]
align_entries = false
reorder_keys = true

```

---

But, later in the trace I can see this (I formatted it with indentation and new line, but it was a single long line):
```
...
TRACE update_configuration:write_request: received response response=Response { 
        jsonrpc: "2.0", 
        id: Number(0), 
        result: Some(Array([
            Object({
                "taplo": Object({
                    "bundled": Bool(true), 
                    "path": Null, 
                    "environment": Object({
                        "RUST_LOG": String("trace")
                    }), 
                    "extraArgs": Array([String("--log-spans"), String("--verbose")]), 
                    "configFile": Object({"path": String("C:\\Users\\PC-Gil\\Documents\\Work\\RustWorkspace\\nova\\taplo.toml"), "enabled": Bool(true)})
                }), 
                "semanticTokens": Bool(false), 
                "schema": Object({
                    "enabled": Bool(true), 
                    "links": Bool(false), 
                    "catalogs": Array([String("https://taplo.tamasfe.dev/schema_index.json"), String("https://www.schemastore.org/api/json/catalog.json")]), 
                    "associations": Object({}), 
                    "cache": Object({"memoryExpiration": Number(60), "diskExpiration": Number(600)})
                    }), 
                "completion": Object({"maxKeys": Number(5)}), 
                "syntax": Object({"semanticTokens": Bool(true)}), 
                "formatter": Object({
                    "alignEntries": Bool(false), 
                    "alignComments": Bool(true), 
                    "arrayTrailingComma": Bool(true), 
                    "arrayAutoExpand": Bool(true), 
                    "arrayAutoCollapse": Bool(true), 
                    "compactArrays": Bool(true), 
                    "compactInlineTables": Bool(false), 
                    "compactEntries": Bool(false), 
                    "columnWidth": Number(100), 
                    "indentTables": Bool(false), 
                    "indentEntries": Bool(false), 
                    "indentString": String("    "), 
                    "trailingNewline": Bool(true), 
                    "reorderKeys": Bool(false), 
                    "allowedBlankLines": Number(2), 
                    "crlf": Bool(false)
                    })
            }), 
            Object({
                "taplo": Object({
                    "bundled": Bool(true), 
                    "path": Null, 
                    "environment": Object({
                        "RUST_LOG": String("trace")
                    }), 
                    "extraArgs": Array([String("--log-spans"), String("--verbose")]), 
                    "configFile": Object({"path": String("C:\\Users\\PC-Gil\\Documents\\Work\\RustWorkspace\\nova\\taplo.toml"), "enabled": Bool(true)})
                }), 
                "semanticTokens": Bool(false), 
                "schema": Object({
                    "enabled": Bool(true), 
                    "links": Bool(false), 
                    "catalogs": Array([String("https://taplo.tamasfe.dev/schema_index.json"), String("https://www.schemastore.org/api/json/catalog.json")]), 
                    "associations": Object({}), 
                    "cache": Object({"memoryExpiration": Number(60), "diskExpiration": Number(600)})
                    }), 
                "completion": Object({"maxKeys": Number(5)}), 
                "syntax": Object({"semanticTokens": Bool(true)}), 
                "formatter": Object({
                    "alignEntries": Bool(false), 
                    "alignComments": Bool(true), 
                    "arrayTrailingComma": Bool(true), 
                    "arrayAutoExpand": Bool(true), 
                    "arrayAutoCollapse": Bool(true), 
                    "compactArrays": Bool(true), 
                    "compactInlineTables": Bool(false), 
                    "compactEntries": Bool(false), 
                    "columnWidth": Number(100), 
                    "indentTables": Bool(false), 
                    "indentEntries": Bool(false), 
                    "indentString": String("    "), 
                    "trailingNewline": Bool(true), 
                    "reorderKeys": Bool(false), 
                    "allowedBlankLines": Number(2), 
                    "crlf": Bool(false)
                    })
            })
        ])
    ), 
    error: None 
} params=Some(ConfigurationParams { 
    items: [
        ConfigurationItem { 
            scope_uri: None, 
            section: Some("evenBetterToml") 
        }, 
        ConfigurationItem { 
            scope_uri: Some(Url { 
                scheme: "file", 
                cannot_be_a_base: false, 
                username: "", 
                password: None, 
                host: None, 
                port: None, 
                path: "/c%3A/Users/PC-Gil/Documents/Work/RustWorkspace/nova", 
                query: None, 
                fragment: None 
            }), 
            section: Some("evenBetterToml") 
        }
    ] 
})

```

And here I see the config I set in `settings.json`. And I can see it twice in the array....
I thing that settings.json overwrite the config in taplo.toml. I am checking that now.

**Comment by tamasfe** [comment]
> I thing that settings.json overwrite the config in taplo.toml.

That shouldn't be the case, it used to be, but I believe I fixed it.

I believe the issue is still what I said unfortunately, if there is no `include` specified in the config, it implicitly uses a glob pattern such as `**/*.toml`, which will not match any files if they use `\`. Can you try setting the following in `taplo.toml`:

```
include = [ "**\\*.toml" ]
```

I have no idea if the glob engine works with it, but it is probably worth a try.

**Comment by GilShoshan94** [comment]
ok, I just commented out the settings in `settings.json` and reload vscode and now the config from `taplo.toml` works.
like that:
```jsonc
    // ## Even Better Toml
    // "evenBetterToml.formatter.alignComments": true,
    // "evenBetterToml.formatter.alignEntries": false,
    // "evenBetterToml.formatter.allowedBlankLines": 2,
    // "evenBetterToml.formatter.arrayAutoCollapse": true,
    // "evenBetterToml.formatter.arrayAutoExpand": true,
    // "evenBetterToml.formatter.arrayTrailingComma": true,
    // "evenBetterToml.formatter.columnWidth": 100,
    // "evenBetterToml.formatter.compactArrays": true,
    // "evenBetterToml.formatter.compactEntries": false,
    // "evenBetterToml.formatter.compactInlineTables": false,
    // "evenBetterToml.formatter.crlf": false,
    // "evenBetterToml.formatter.indentEntries": false,
    // "evenBetterToml.formatter.indentString": "    ",
    // "evenBetterToml.formatter.indentTables": false,
    // "evenBetterToml.formatter.reorderKeys": false,
    // "evenBetterToml.formatter.trailingNewline": true,
    "evenBetterToml.schema.enabled": true,
    "evenBetterToml.schema.links": false,
    "evenBetterToml.semanticTokens": false,
    "evenBetterToml.syntax.semanticTokens": true,
    "evenBetterToml.taplo.configFile.path": "C:\\Users\\PC-Gil\\Documents\\Work\\RustWorkspace\\nova\\taplo.toml",
    "evenBetterToml.taplo.environment": {
        "RUST_LOG": "trace"
    },
    "evenBetterToml.taplo.extraArgs": [
        "--log-spans",
        "--verbose"
    ],
```

---
## My conclusions:
1. The config in `settings.json` is overwriting the local config `taplo.toml`.
It seems to me that the desired behavior is the oposite, that a local config in a project would take precedence over  the ide config. In any case this behavior needs to be documented for the users.

2. The path for `"evenBetterToml.taplo.configFile.path"` is not fully working, by that I mean that an absolute path works, but not a relative one, at least on Windows.
Also, I just tried with Unix path and it works as well:
```json
"evenBetterToml.taplo.configFile.path": "C:/Users/PC-Gil/Documents/Work/RustWorkspace/nova/taplo.toml",
```
3. The config file `taplo.toml` or `.taplo.toml` is not found automatically unless explicitly set in "evenBetterToml.taplo.configFile.path". At least on Windows.

**Comment by GilShoshan94** [comment]
> > I thing that settings.json overwrite the config in taplo.toml.
> 
> That shouldn't be the case, it used to be, but I believe I fixed it.
> 
> I believe the issue is still what I said unfortunately, if there is no `include` specified in the config, it implicitly uses a glob pattern such as `**/*.toml`, which will not match any files if they use `\`. Can you try setting the following in `taplo.toml`:
> 
> ```
> include = [ "**\\*.toml" ]
> ```
> 
> I have no idea if the glob engine works with it, but it is probably worth a try.

I just gave it a try, by uncommenting my settings.json and by adding the `include = [ "**\\*.toml" ]` to the `taplo.toml` file.
It does not work.

**Comment by tamasfe** [comment]
Thank you for the thorough dive in!

I've fixed the formatting options where the priority was indeed still incorrect.

2. and 3. should be fixed by #285, I have no ETA on that unfortunately.

**Comment by GilShoshan94** [comment]
Happy I could help a little and contribute !

In the meantime for 2. and 3. there is the workaround to explicitly set the full absolute path.

I am coding in Rust, but it sounds like this issue is in the JS side (which I don't really know, nor TypeScript).
If you guide me/point me where to look for the paths issues I can look into it.

I am closing this issue since you fixed 1., 2. and 3. is in #285 and the hot reload thing is #282

Thank you !
